src/math/power.c

```diff
@@ -0,0 +1,78 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2021 Intel Corporation. All rights reserved.
+//
+// Author: Shriram Shastry <malladi.sastry@linux.intel.com>
+//
+//
+
+ /* Include Files */
+#include <sof/math/power.h>
+
+/* Function Definitions */
+/**
+ * Arguments	: int64_t denominator
+ *
+ * Return Type	: int32_t
+ */
+static inline int32_t div_s32s64(int64_t denominator)
+{
+	uint64_t b_denominator;
+	int32_t quotient;
+
+	if (denominator == 0) {
+		quotient = INT32_MAX;
+	} else {
+		b_denominator =
+		    (denominator < 0) ? ~(uint64_t)denominator + 1 : (uint64_t)denominator;
+		b_denominator = POWER_B_NUMERATOR / b_denominator;
+
+		if (denominator < 0)
+			/* for negative denominator, typecast uint64_t to int32_t for */
+			/* negative quotient */
+			quotient = (int32_t)-(int64_t)b_denominator;
+		else
+			quotient = (int32_t)b_denominator;
+	}
+	return quotient;
```

Comment:
I wonder why this function is so complex, it just divides 2^50 by parameter denominator, right? Signed 64 bit divide should work in C.

Reply:
I tried myself with Xplorer, there really seems to be a problem in xt-xcc with simple test function like. It calculates wrong for example for `num = 1125899906842624LL;` and `den = -1099511627776LL;`

```
int64_t int_div(int64_t num, int64_t den)
{
	return num / den;
}
```
I think it's good idea to add your divide function comment  of issue why you use such workaround for 64 bit division.

Reply:
_`it just divides 2^50 by parameter denominator, right ?`_
That's correct 

Un-optimized version i.e. my initial version snippet where I didn't use 2^50 numerator constant and support XCC compilation , accuracy 

```
static int32_t div_s32s64(int64_t numerator, int64_t denominator)
{
        uint64_t b_denominator;
        uint64_t b_numerator;
        int32_t quotient;
        if (denominator == 0LL) {
                if (numerator >= 0LL) {
                        quotient = INT32_MAX;
                } else {
                        quotient = INT32_MIN;
                }
        } else {
                if (numerator < 0LL) {
                        b_numerator = ~(uint64_t)numerator + 1ULL;
                } else {
                        b_numerator = (uint64_t)numerator;
                }
                if (denominator < 0LL) {
                        b_denominator = ~(uint64_t)denominator + 1ULL;
                } else {
                        b_denominator = (uint64_t)denominator;
                }
                b_numerator /= b_denominator;
                if ((numerator < 0LL) != (denominator < 0LL)) {
                        quotient = (int32_t) - (int64_t)b_numerator;
                } else {
                        quotient = (int32_t)b_numerator;
                }
        }
        return quotient;
}

```

Reply:
@lyakh has seen similar thing to happen. I tried in my above function to print the value q = num / den before return and it was correct! The corruption of int64_t value happens in return. So, in theory you should be able to get a simple divide code to work if you don't use a function for it.

Reply:
The possibility lies in when denominator or numerator is smaller than zero case , I am doing typecast `int64` to `int32`. I have added comment https://github.com/thesofproject/sof/pull/4941/files#r741981730